### DISCUSSION TITLE
Raspberry Pi, Roland RCC-10-USXF USB Microphone cable, monitoring on pHAT DAC

### DISCUSSION THREAD

**Discussion by dubnemo** [jamulussoftware/jamulus#1376] [Technical Support & Help]
I can get the Roland RCC-10-USXF USB Microphone to route to the Jamulus input, but jack seems to not allow me to route the Jamulus output to the Raspberry Pi Headphones.  
It seems that Jamulus needs to talk to ALSA directly to do this.  I can do this perfectly fine in Audacity without Jack.

The OBVIOUS goals are 1) simplification and 2) reduced latency.

Has anyone got around this?

![2021-03-28-085517_1920x1080_scrot](https://user-images.githubusercontent.com/32434235/112754813-73318b00-8fa3-11eb-9ad7-babcd3a2e9da.png)

**Comment by softins** [comment]
Have you tried the **Advanced** tab in QJackCtl? It offers you independent setting of input and output audio interfaces.

**Reply by dubnemo** [reply]
Thank you @softins, getting somewhere with that...appreciate that as it does allow me to select other devices.  I hear crackles but I am over wi-fi at this moment, and my server is not 3.7 yet.  Is Jamulus capable of 16 bit?  I seem to recall reading 24 bit.

![2021-03-28-220324_1920x1080_scrot](https://user-images.githubusercontent.com/32434235/112782104-df4ad800-9011-11eb-899c-ac012e1a0a9a.png)



Cannot connect to server request channel
jackdmp 1.9.12
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2017 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
audio_reservation_init
Acquire audio card Audio2
Acquire audio card Audio1
creating alsa driver ... hw:Headphones,0|hw:Device,0|1024|2|48000|0|0|nomon|swmeter|-|32bit
configuring for 48000Hz, period = 1024 frames (21.3 ms), buffer = 2 periods
ALSA: final selected sample format for capture: **16bit little-endian**
ALSA: use 2 periods for capture
ALSA: final selected sample format for playback: **16bit little-endian**
ALSA: use 2 periods for playback
qt5ct: D-Bus global menu: no
QFSFileEngine::open: No file name specified

**Reply by jujudusud** [reply]
Set in settings tab of Qjackctl :
Frames/period : 128
Periods/buffer : 3 (better for usb device)
In Jamulus :
Do note use  "enable small network buffer"

**Comment by dubnemo** [comment]
I have the other RPi4 working, but trying to build a RPi3B for my studio.  
I cannot get Jamulus 3.7 to connect to Jackd as a client with the current build instructions that include QJackCtl.
https://jamulus.io/wiki/Installation-for-Linux

Replicated on THREE separate RPis devices and three FRESH microSD card images.

I am so tired of Jack. Why cannot  we just use ALSA on Raspberry Pi OS?

**Reply by dubnemo** [reply]
Attempt number #4, this time without QJackCtl install.  
https://jamulus.io/wiki/Installation-for-Linux
sudo apt-get install build-essential qt5-qmake qtdeclarative5-dev qt5-default qttools5-dev-tools **libjack-jackd2-dev**

Jamulus not starting Jackd, 

![2021-04-29-074005_1824x984_scrot](https://user-images.githubusercontent.com/32434235/116555012-7df08000-a8c1-11eb-81b7-6ba5347123da.png)

**Reply by jujudusud** [reply]
You have to start jack first.
Which version Do you have installed ? Jack2 ?

**Reply by dubnemo** [reply]
It is in the apt-get install build-essential command already.  libjack-jackd2-dev

**Reply by dubnemo** [reply]
Jamulus use to start without me having to do manually start Jackd.

**Reply by jujudusud** [reply]
No, that's wrong and it is written in the Jamulus windows :
"Try to start Jack server manually".
Qjackctl starts Jackd2 for you certainly.
libjack-jackd2-dev is a library to build Jamulus code if you want to buil by yourself. It's not the server...
You can start Jack server manually with the following command (you have to work around the values):
jackd -P70 -p16 -t2000 -dalsa -dhw:0 -p128 -n3 -r48000 -s &

Please read my comment here : https://github.com/jamulussoftware/jamulus/discussions/1376#discussioncomment-673779

**Reply by hoffie** [reply]
> Jamulus use to start without me having to do manually start Jackd.

Yes, that's also my experience -- if no jackd is running, Jamulus will implicitly start one.

**Reply by jujudusud** [reply]
Only if d-bus is used?

**Reply by hoffie** [reply]
Ah, this could be, yes.

**Reply by jujudusud** [reply]
Certainly uninstalled at the same time as Qjackctl... :-(

**Comment by jujudusud** [comment]
the problem is related here also : https://github.com/jackaudio/jack2/issues/691
High latency with Jack2 and Raspberry Pi 4 (snd_bcm2835).
You should have a look to a usb audio interface or Pisound hat : https://blokas.io/pisound/

**Reply by jujudusud** [reply]
The integrated sound chipset on Rpi 4 is 44.1kHz and 16 bits. If you want to use it, you have to resample the Jack output with zita_j2a to 44.1kHz to hear the sound on your headset.

**Reply by dubnemo** [reply]
@jujudusud the resampling point is well taken.  And that could cause issues such as latency that may defeat the purpose of the experiment.  Thank you.

**Reply by jujudusud** [reply]
If you do not have a 48kHz device, you can't use Jamulus without resampling.

**Reply by dubnemo** [reply]
@jujudusud understand that, hence my latency point.  

I have a HiFi Berry DAC + ADC Pro for recording on a RPi4, but its RCA output had to be routed to a mixer/amp. https://www.hifiberry.com/shop/boards/hifiberry-dac-adc-pro/

Have you actually used the Pisound?  Curious if the 1/4" output can drive headphones, or whether I have to use an external mixer/amp?  The spec of 0 to 2.1V doesn't give me info of drive/ current draw.

The Roland USB microphone cable would have been a simple low-latency RPi3B solution for a lot of users if the RPi audio chipset could support 48Khz.   I will see what 48khz -> 44.1khz conversion adds to latency, but I speculate it would throw it RED - once I find a branch that starts jackd again ;-).

**Reply by jujudusud** [reply]
I didn't use Pisound with Jamulus.
Only tested with modep to have an audio/midi multieffect with deported webui.
If I have Time to, I will have a look to that.

**Comment by jujudusud** [comment]
Alors ?
Is your Raspberry Pi 4 set up properly ?

**Comment by dubnemo** [comment]
Here is proof that Jamulus normally starts jackd by default (this is an older installation on RPi4 that I use at my home, not at my studio (different properties):

not running ps -ef
![2021-05-02-073431_1920x1080_scrot](https://user-images.githubusercontent.com/32434235/116813613-2ac53a00-ab1a-11eb-8ee7-65c76690df6e.png)

Jamulus running, jackd now running
![2021-05-02-073533_1920x1080_scrot](https://user-images.githubusercontent.com/32434235/116813630-39135600-ab1a-11eb-9907-f8f9ecfe9f89.png)

Jamulus closed, jackd no longer running.
![2021-05-02-073551_1920x1080_scrot](https://user-images.githubusercontent.com/32434235/116813642-4d575300-ab1a-11eb-800c-a4f4032e1ad9.png)

@jujudusud as stated, FRESH OS install. FRESH Jamulus build.  Roland device is USB, and now I cannot even build a new machine to use in my studio.  The point of this project is to minimize latency, using internal headphones, USB mic cable - I have numerous I/O devices (don't go there).   I created a separate code repo issue a long while ago that Jamulus should support ALSA (which would have SOLVED) this, but it was hi-jacked to for Portaudio.   Perhaps that will work, as I don't care jack about Jack. ALSA is used by Jackd and adds overhead.

**Reply by jujudusud** [reply]
In your last exemple, Jamulus can use "JACK" because d-bus is in use. If Jamulus can't open JACK there is a problem with d-bus on your machine.
As you can see on the last picture, JACK and ALSA can't see your "device" selected to capture. There is a problem here with your command line to start jackd.

**Reply by dubnemo** [reply]
@jujudusud  I didn't have it plugged into the RPi4.  Just showing that I shouldn't to start jackd manually.  I will research more on older Jamulus branches for my RPi3B, as the latest is not starting jackd.   The one that is working is the RPi4, running a dev branch of 3.7.0.

**Reply by jujudusud** [reply]
As you wanted.

**Comment by dubnemo** [comment]
My direction to use Roland RCC-10-USXF interface cable is to use a pHAT DAC unit for the headphone output.  I realized I had in my inventory a pimoroni PIM200 pHAT DAC (now discontinued) that uses the same chipset as the HiFiBerry, leverages a similar ALSA asound.conf setup and overlay as the HiFiBerry.  These DAC are around $15 USD and require soldering.  
https://learn.pimoroni.com/tutorial/phat/raspberry-pi-phat-dac-install
Total hardware cost $5 (Roland + pHAT DAC).
Firing up the soldering gun this Sunday morning at the Studio.

Then find the Jamulus branch for portaudio work, and see I can say Good-bye to Jackd for reduced overhead.
With ADC in the Roland right at the USB connector, I anticipate additoinal reduced latency due to elimination of USB cable length; e.g., my Antelope Goliath at 15ft has typical overall 70ms even with a Jamulus server on the same network @2ms ping, whereas Behringer UMD204HD at 3ft has 35ms to the same server.
The horsepower of the RPi3B will be a factor too.
It will then be closer to the jamlink hardware I have used for nearly six years -- which Jamulus still cannot touch performance-wise.

**Reply by npostavs** [reply]
I recently updated the portaudio branch to work on Linux, by the way. I did successfully get sound through the ALSA backend. #870

**Reply by dubnemo** [reply]
> I recently updated the portaudio branch to work on Linux, by the way. I did successfully get sound through the ALSA backend. #870

@npostavs Noam that is great news.  I will await the release.  I do have the above working with Jackd, finally, but it is not the performance expected.   Hope to provide the before and after total latency numbers.

**Reply by jujudusud** [reply]
JACK is not adding latency. With or without your latency problem (if resolvable) is elsewhere.

**Reply by dcorson-ticino-com** [reply]
> I anticipate additoinal reduced latency due to elimination of USB cable length

The difference between ~5 meters and ~1 meter cable length is not why one interface is slower than the other.
Propagation delay in a copper signal cable is around 5ns (0.000005ms)  a meter. 
That difference will be the difference in buffering, processing and communication of the interfaces.

**Reply by gene96817** [reply]
It is also. possible to have some USB problems if you have enough USB cables and more than one. USB hub. If you have just one hub and most of your USB cables are short, then there shouldn't be a problem. To most people, the USB is either working or not. However if some USB problems are developing, the impact on latency in the scale of a few tens of microseconds (us, where 1000us = 1ms). For the most part, USB bus problems are not the issue.

**Reply by dubnemo** [reply]
> JACK is not adding latency. With or without your latency problem (if resolvable) is elsewhere.

In issue https://github.com/jamulussoftware/jamulus/issues/116 for the jamulus repo, I requested direct ALSA support.  There were a number of supporters for this for example,  "So, as long as you have only one application using sound running, ALSA has much less overhead (less processes ⇒ less RAM and less context switches)."  This issue was switched to PortAudio for 'compatibiity' reasons which disappointed me.  

So unless you know the 'expense' of programming threads, please do not comment.  It is about reducing the number of threads and therefore the efficiency of the entire 'system'.

As far as the cable length, that too is ridiculous.  There are maximun cable length specifications for data communications reasons.  The signal degrades with length for multitude of reasons.  If you take an electrical engineering course on digital communications (yours truly has), you will start to understand the reasons, and there are multiple reasons.  It is not related to theoretical speed of materials, but Schrödinger wave properties, quality of connectors, and the quality of the algorithms used to encode and decode the signal.

**Reply by jujudusud** [reply]
@dubnemo Your problems with your issue 116, that I don't remember the existence before you told me here, are not part of this discussion.
If you don't want to have answers on a public discussion, please don't ask ! Thanks a lot.
For your problem with the issue 116, please report in this issue and not here as we cannot understand what you are talking about.

That said, in this discussion, you are not talking before this message about reducing the number off threads on your system.
So ok, you want to see Jamulus using ALSA. No one here has started this job. Please code and provide a PR.

**Reply by dubnemo** [reply]
GREAT, I have looked but no Getting Started for Developers wiki.  Is that hidden?

**Reply by dubnemo** [reply]
And yes, it DID evolve back to that topic.  But I am trying to squeeze every option I have to improve performance, specifically with the phatDAC and Roland USB mic cable.  Reduce everything I can, and cheaply.
